src/metabase/lib/schema/expression.cljc

```diff
@@ -1,39 +1,181 @@
+   ;; for some crazy reason `:type/Float` is the common base type of all numbers with non-integer-real places, not something
+   ;; smart like `:type/Decimal`
+   (ref-of-base-type :type/Float)])
 
 ;;; Any expression that returns any kind of number.
 (mr/def ::number
   [:or
+   {:error/message "expression returning a number"}
    ::integer
-   ::floating-point])
+   ::non-integer-real])
+
+;;; `:datetime-add` with a Date expression:
+;;;
+;;;    [:datetime-add <options> <date-expression> <amount> <date-interval-unit>]
+(mr/def ::datetime-add.date
+  [:schema
+   [:tuple
+    [:= :datetime-add]
+    ::common/options
+    #_expression [:ref ::date]
+    #_amount     [:ref ::integer]
+    #_unit       ::temporal-bucketing/unit.date.interval]])
+
+;;; Any expression that returns a `:type/Date`
+(mr/def ::date
+  [:or
+   {:error/message "expression returning a date"}
+   [:ref ::literal/date]
+   [:ref ::datetime-add.date]
+   (ref-of-base-type-with-temporal-unit :type/Date [:not ::temporal-bucketing/unit.date.extract])
+   ;; TODO -- does a truncation of a :type/DateTime to `::temporal-bucketing/unit.date.truncate` like `:day` return a
+   ;; `:type/Date`? Should we act like it does?
+   #_(ref-of-base-type-with-temporal-unit :type/DateTime ::temporal-bucketing/unit.date.truncate)])
+
+;;; `:datetime-add` with a Time expression:
+;;;
+;;;    [:datetime-add <options> <time-expression> <amount> <time-interval-unit>]
+(mr/def ::datetime-add.time
+  [:schema
+   [:tuple
+    [:= :datetime-add]
+    ::common/options
+    #_expression [:ref ::time]
+    #_amount     [:ref ::integer]
+    #_unit       ::temporal-bucketing/unit.time.interval]])
+
+;;; Any expression that returns a `:type/Time`
+(mr/def ::time
+  [:or
+   {:error/message "expression returning a time"}
+   [:ref ::literal/time]
+   [:ref ::datetime-add.time]
+   (ref-of-base-type-with-temporal-unit :type/Time [:not ::temporal-bucketing/unit.time.extract])])
+
+;;; `:datetime-add` with a DateTime expression:
+;;;
+;;;    [:datetime-add <options> <date-time-expression> <amount> <date-time-interval-unit>]
+(mr/def ::datetime-add.date-time
+  [:tuple
+   {:error/message ":datetime-add clause with a date time expression"}
+   [:= :datetime-add]
+   ::common/options
+   #_expression [:ref ::date-time]
+   #_amount     [:ref ::integer]
+   #_unit       ::temporal-bucketing/unit.date-time.interval])
+
+;;; Any expression that returns a `:type/DateTime`
+(mr/def ::date-time
+  [:or
+   {:error/message "expression returning a date time"}
+   [:ref ::literal/date-time]
+   [:ref ::datetime-add.date-time]
+   (ref-of-base-type-with-temporal-unit :type/DateTime [:not ::temporal-bucketing/unit.date-time.extract])])
 
 ;;; Any expression that returns some sort of temporal value `java.time.OffsetDateTime`
 (mr/def ::temporal
-  ;; TODO
   [:or
-   ::literal/temporal])
+   {:error/message "expression returning a date, time, or date time"}
+   ::date
+   ::time
+   ::date-time])
```

Comment:
I split out `::temporal` expression types into three separate subtypes, `::date`, `::time`, and `::date-time`. The distinction is important, because there are things we support that don't make sense for every temporal type, e.g. you cannot extract an `:hour` from a `::time`